### DISCUSSION TITLE
Supporting both .Net Core and .Net Framework in one process.

### DISCUSSION THREAD

**Discussion by DmitrySyrkin** [agracio/edge-js#194] [Q&A] [answered]
Good morning, everyone.
First of all, thanks to @agracio for the work he is doing.

My question is as follows:
we are using edge-js with Specflow tests and now are are switching our backend API from .Net Framework 4.8 to .Net 8.0.
The main issue, as far as I understand, that prevents me to use edge-cs for both new and old versions of API is having all settings as environment variables. For example, EDGE_APP_ROOT should be empty in case of .net framework and should be set for .net core.

I can imagine rebuilding edge-js from source code with settings renaming (like EDGE_CORE_APP_ROOT) and referencing two version of package, but, not sure it is a great idea.
 
 If anyone can share thoughts\experience, I'd really appreciate that. 
 
 Thank you. 
 Dmitry Syrkin

**Comment by agracio** [comment]
This is actually not an issue when switching between .NET Core  and .NET Framework. You can always set `EDGE_APP_ROOT` variable but .NET Framework ignores it, you can see this use case in  `tools/test.js`. 

However when switching between .NET Core and .NET Standard it is important to NOT set `EDGE_APP_ROOT` for .NET Standard. You can see that in[ https://github.com/agracio/edge-js-quick-start `main.js`](https://github.com/agracio/edge-js-quick-start/blob/2d0203a08a1e4fb65d0498835252b3933c679665/main.js#L10). But if you have such a use case then consider sending runtime parameters to your process.

**Comment by agracio** [comment]
Another important thing to clarify is usage of different .NET flavours in **one process** as you put it in the title of this discussion. It is not possible to switch between .NET Framework and .NET Core during the process lifetime once Edge.Js is loaded. 

For example setting `EDGE_USE_CORECLR=1` running some code and then setting `EDGE_USE_CORECLR=undefined` programmatically from within the running application then trying to run some code will lead to unexpected Edge.js errors for second .NET flavour.

This does not mean that `EDGE_USE_CORECLR` variable has to be specified via command line, but when running .NET Core application it needs to be set before executing 'const edge = require('edge-js');'.
```js
process.env.EDGE_USE_CORECLR = 1;
const edge = require('edge-js');
```

**Reply by DmitrySyrkin** [reply]
Thank you very much for quick response. I've rechecked if we used netstandard, but looks like you've spotted the real problem in your comment. So, for now I see two possibilities: 1) try to rebuild your project with different env settings names or 2) try to migrate our application to .net core.

**Reply by agracio** [reply, marked as answer]
Not sure if I understand your reply correctly, if you use .NET Standard then you don't need to set `EDGE_APP_ROOT` parameter at all. 
And if you use .NET Core then you can set `EDGE_APP_ROOT` and it will be ignored by .NET Framework and only used by .NET Core. 
I don't see what benefit you can get from adding extra parameter in Edge.Js. 

Here is an explanation of `EDGE_APP_ROOT` usage.

| Framework      | `EDGE_APP_ROOT` Impact                      |
|-----------------|---------------------------------------------|
| .NET 4.x           | Ignored, can be set to anything              |
| .NET Core        | Required   if dll paths are not absolute   |
| .NET Standard | Should NOT be set                                  |

**Reply by agracio** [reply]
Let me know if the comment above clarifies your question.

**Reply by DmitrySyrkin** [reply]
Thank you very much for clarifications. We have some netstandard libraries in a solution, but the one we use for edge-js is on .net framework (that's why I was a bit confused).
My current understanding is that not EDGE_APP_ROOT  is the issue but EDGE_USE_CORECLR itself. And we cannot set it to 1 and use with both .net8 and net48 projects in same process (e.g. because of using HttpContext from System.Web). I'll try to discuss with the team if migration to net8 can be done.
Please let me know if I'm missing anything again and thank you for your time and support.

**Reply by agracio** [reply]
You can run .NET Core and .NET Standard applications without setting `EDGE_USE_CORECLR` variable. 
Take https://github.com/agracio/edge-js-quick-start as an example, it will run in both Core and Standard mode even if `EDGE_USE_CORECLR` not set, the underlying code will default to using .NET Framework execution and the code will run just fine. 

Also while testing different variations of the code I discovered that  `EDGE_APP_ROOT` is not actually required if your specify absolute paths to dlls that you are using. 
For example in https://github.com/agracio/edge-js-quick-start I specify absolute paths to `baseDll` parameters in each method , in this case  `EDGE_APP_ROOT` is not required.

```js
var getAppDomainDirectory = edge.func({
    assemblyFile: baseDll,
    typeName: localTypeName,
    methodName: 'GetAppDomainDirectory'
});
```
 
But in `edge-js` tests `EDGE_APP_ROOT` is set while `assemblyFile` is passed a name `test` without .dll extension. This tell Edje.Js to look for that assembly in the specified root.

**Reply by DmitrySyrkin** [reply]
I think that when I tried to load net8.0 assembly without CORECLR flag set, I was getting
`Could not load file or assembly 'System.Runtime, Version=8.0.0.0, `

I'll recheck with quick-start project and will let you know. Thank you!

**Reply by agracio** [reply]
Have you updated to the latest Edge.Js release? 
For me https://github.com/agracio/edge-js-quick-start runs fine with or without setting `EDGE_USE_CORECLR` flag.

**Reply by DmitrySyrkin** [reply]
edge-js 21.7.2
npm 10.1.0
node 18.18.0

So, if I comment this line: https://github.com/agracio/edge-js-quick-start/blob/master/main.js#L9
I'm getting 


> ...\edge-js-quick-start\node_modules\edge-js\lib\edge.js:178
> return edge.initializeClrFunc(options);
>                      ^
> Error: Could not load file or assembly 'System.Runtime, Version=7.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.

**Reply by agracio** [reply]
Very strange I am not getting that exception on my PC with `npm start` or `npm run start:standard`. What .NET version do you have installed `dotnet --version`?

**Reply by DmitrySyrkin** [reply]
8.0.200

I've tried to do the following: updated all cpp files changing EDGE_ to EDGE_CORE_ for env variables, managed to build the project locally (thank you for the /tools folder) and update main.js.
No luck - still "Could not load file or assembly 'System.Runtime, Version=8.0.0.0..."  (I've rebuilt QuickStart.Core with 8.0)
I'll try to rebuild the solution without any changes and check if it works with default main.js right now.

Once again, thank you for your time!

UPD: I found that some of env variables are also used in related packages (e.g. edge-cs and Edge.Js.Csharp). 
I've tried to rebuild them also with no luck.
I'll try to rebuild edge_nativeclr with new flags tomorrow.

Just to check - you don't have any environmental variables set on user/system level that can affect your local testing?

**Reply by agracio** [reply]
You are correct, I was reusing the same CMD window that had `EDGE_USE_CORECLR` set and now I am getting the same error.  What is your use case for this, you have .NET Framework solution where some of the projects are migrating to .NET Standard/Core or are you migrating the entire solution?

**Reply by DmitrySyrkin** [reply]
Long story short: we use egde-js for Specflow tests. Now we are in process of migrating C#-api from net48 to net8.0 but for some decent period of time we need to support both. We are using typescript and for some historical reasons do not reference edge-js package but have a local copy of it with some minor tweaks  (some _nodeRequire calls). So, I planned to add one more local copy of edge-js that will use different set of environment variables but (being a humble C# developer) have not managed to rebuild everything (edge-js, edge-cs, Edge.Js.CSharp, anything else) from the first try. Maybe the way is bad from the beginning and some other approach should be used.

If you have any ideas, I'll really appreciate any help and guidance. Thank you.

**Reply by DmitrySyrkin** [reply]
Just wanted to share about intermediate success. Temporary solution is as follows:
1) two local copies to edge-js (renamed to edge2-js and edge3-js)
2) first one left as is, latter: environment settings - added EDGE_CORE_ instead of EDGE_, edge_coreclr.node  - rebuilt with removing USE_CORECLR checks (I need it to be treated as always true).

Next step is fix working with EDGE_APP_ROOT, now I can use only one net8.0 library with no other references.

Thank you very much, I will mark discussion as Answered, but if you have anything to be added (or plans for next versions) - that would be great.

**Reply by agracio** [reply]
It is still hard for me to follow so lets start from the beginning but to me it looks like you are overthinking it.
First I have amended the table above that you marked as answer:

| Framework      | `EDGE_APP_ROOT` Impact                      |
|-----------------|---------------------------------------------|
| .NET 4.x           | Ignored, can be set to anything              |
| .NET Core        | Required if dll paths are not absolute   |
| .NET Standard | Should NOT be set                                  |

This amendment means that that  `EDGE_APP_ROOT` is only required if your code does not specify absolute path to dlls that you are using like it is done in quick-start. In case of quick-start `EDGE_APP_ROOT` is not required and can be removed. Same goes for any inline C# code that you have, `EDGE_APP_ROOT` is not needed.

To explain further the following code in quick-start has `baseDll` parameter set to an absolute path including .dll extension. In my case it is ` C:\Users\myusername\Dev\edge-js-quick-start\src\QuickStart.Core\bin\Debug\net7.0\QuickStart.Core.dll`. Using such code does not require `EDGE_APP_ROOT`. 
However I could also specify `baseDll` parameter as `QuickStart.Core` this time without .dll extension then `EDGE_APP_ROOT` needs to be set to `C:\Users\myusername\Dev\edge-js-quick-start\src\QuickStart.Core\bin\Debug\net7.0` for the code to find it.

```js
var getAppDomainDirectory = edge.func({
    assemblyFile: baseDll,
    typeName: localTypeName,
    methodName: 'GetAppDomainDirectory'
});
```

Now lets talk about your specific case, to me it makes no sense to modify Edge.Js in the first place. As I already mentioned specifying  `EDGE_APP_ROOT` does not interfere with your .NET Framework project so you can use it anyway so your only problem is  `EDGE_USE_CORECLR`

I can only assume that while you are migrating you are running 2 sets of test suites one for old net48 project and one for project undergoing migration. Can you not run one with `EDGE_USE_CORECLR` set and another one when its not set. 

I even run both .NET Framework and .NET Core tests in a single AppVeyor job you can take a look at https://github.com/agracio/edge-js/blob/master/appveyor.yml for reference lines 30 and 32. And here is an example of the test results https://ci.appveyor.com/project/agracio/edge-js/branch/master/job/2tkpsnxxdyvcxip0/tests

Perhaps you could explain your situation with tests better and we could come up with a more streamlined solution.

**Reply by DmitrySyrkin** [reply]
Quick reply: it is one process (you "open" the html with selenium that has lots of calls for both new and old api. It is not so easy to split them to separate processes.

I guess your ".net framework application" can be run using coreclr. You can try to add reference to System.Web and use smth like HttpContext in code. I believe you can do it with  edge-js-quick-start (i've  just changed netstandard to net48 and added a reference)


And sorry for confusing: EDGE_USE_CORECLR and EDGE_APP_ROOT - are different issues. So, I'm try to "fix" them one after another.

**Reply by agracio** [reply]
I have serious doubts that you will be able to run both .NET Framework and .NET Core/Standard in one process. Depending on `EDGE_USE_CORECLR` tag Edge.Js will bind itself to completely different code paths both in C# and C++.
